src/coreclr/jit/fgbasic.cpp

```diff
@@ -2191,35 +2191,37 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, Fixed
 
 void Compiler::fgAdjustForAddressExposedOrWrittenThis()
 {
+    LclVarDsc* thisVarDsc = lvaGetDesc(info.compThisArg);
+
     // Optionally enable adjustment during stress.
     if (!tiVerificationNeeded && compStressCompile(STRESS_GENERIC_VARN, 15))
     {
-        lvaTable[info.compThisArg].lvHasILStoreOp = true;
+        thisVarDsc->lvHasILStoreOp = true;
     }
 
     // If this is exposed or written to, create a temp for the modifiable this
-    if (lvaTable[info.compThisArg].lvAddrExposed || lvaTable[info.compThisArg].lvHasILStoreOp)
+    if (thisVarDsc->IsAddressExposed() || thisVarDsc->lvHasILStoreOp)
     {
         // If there is a "ldarga 0" or "starg 0", grab and use the temp.
         lvaArg0Var = lvaGrabTemp(false DEBUGARG("Address-exposed, or written this pointer"));
         noway_assert(lvaArg0Var > (unsigned)info.compThisArg);
-        lvaTable[lvaArg0Var].lvType            = lvaTable[info.compThisArg].TypeGet();
-        lvaTable[lvaArg0Var].lvAddrExposed     = lvaTable[info.compThisArg].lvAddrExposed;
-        lvaTable[lvaArg0Var].lvDoNotEnregister = lvaTable[info.compThisArg].lvDoNotEnregister;
+        LclVarDsc* arg0varDsc = lvaGetDesc(lvaArg0Var);
+        arg0varDsc->lvType    = thisVarDsc->TypeGet();
+        arg0varDsc->SetAddressExposed(thisVarDsc->IsAddressExposed() DEBUGARG(thisVarDsc->GetAddrExposedReason()));
+        arg0varDsc->lvDoNotEnregister = thisVarDsc->lvDoNotEnregister;
 #ifdef DEBUG
-        lvaTable[lvaArg0Var].lvVMNeedsStackAddr = lvaTable[info.compThisArg].lvVMNeedsStackAddr;
-        lvaTable[lvaArg0Var].lvLiveInOutOfHndlr = lvaTable[info.compThisArg].lvLiveInOutOfHndlr;
-        lvaTable[lvaArg0Var].lvLclFieldExpr     = lvaTable[info.compThisArg].lvLclFieldExpr;
-        lvaTable[lvaArg0Var].lvLiveAcrossUCall  = lvaTable[info.compThisArg].lvLiveAcrossUCall;
+        arg0varDsc->SetDoNotEnregReason(thisVarDsc->GetDoNotEnregReason());
 #endif
-        lvaTable[lvaArg0Var].lvHasILStoreOp = lvaTable[info.compThisArg].lvHasILStoreOp;
-        lvaTable[lvaArg0Var].lvVerTypeInfo  = lvaTable[info.compThisArg].lvVerTypeInfo;
+        arg0varDsc->lvHasILStoreOp = thisVarDsc->lvHasILStoreOp;
+        arg0varDsc->lvVerTypeInfo  = thisVarDsc->lvVerTypeInfo;
 
         // Clear the TI_FLAG_THIS_PTR in the original 'this' pointer.
-        noway_assert(lvaTable[lvaArg0Var].lvVerTypeInfo.IsThisPtr());
-        lvaTable[info.compThisArg].lvVerTypeInfo.ClearThisPtr();
-        lvaTable[info.compThisArg].lvAddrExposed  = false;
-        lvaTable[info.compThisArg].lvHasILStoreOp = false;
+        noway_assert(arg0varDsc->lvVerTypeInfo.IsThisPtr());
+        thisVarDsc->lvVerTypeInfo.ClearThisPtr();
+        // Note that here we don't clear `m_doNotEnregReason` and it stays
```

Comment:
Why don't we clear address exposed reason?

Reply:
I saw only 1 method that was using this logic in benchmark collection so I did not dig deep here.

In `fgAdjustForAddressExposedOrWrittenThis` we are creating a clone for `this` and replace other uses of the original `this` with the clone, so we assume that now the reason why `this` was marked as address exposed is no longer valid and there could be no other reasons why it could be address exposed.  
So after this transformation, the clone is address exposed and the original - not.

However, it is questionable if we can clear `lvDoNotEnregister` here as well because we don't have information if `AddrExposed` was the only reason why we can't enregister. So we can imagine a situation where we clear this flag and enregister the original `this` when we should not, for example, because local enreg is disabled or something like that.